core/src/main/java/org/apache/iceberg/formats/FormatModelRegistry.java

```diff
@@ -0,0 +1,234 @@
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A registry that manages file-format-specific readers and writers through a unified object model
+ * factory interface.
+ *
+ * <p>This registry provides access to {@link ReadBuilder}s for data consumption and various writer
+ * builders:
+ *
+ * <ul>
+ *   <li>{@link WriteBuilder} for basic file writing,
+ *   <li>{@link DataWriteBuilder} for data files,
+ *   <li>{@link EqualityDeleteWriteBuilder} for equality deletes,
+ *   <li>{@link PositionDeleteWriteBuilder} for position deletes.
+ * </ul>
+ *
+ * The appropriate builder is selected based on {@link FileFormat} and object model name.
+ *
+ * <p>{@link FormatModel} objects are registered through {@link #register(FormatModel)} and used for
+ * creating readers and writers. Read builders are returned directly from the factory. Write
+ * builders may be wrapped in specialized content file writer implementations depending on the
+ * requested builder type.
+ */
+public final class FormatModelRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(FormatModelRegistry.class);
+  // The list of classes which are used for registering the reader and writer builders
+  private static final List<String> CLASSES_TO_REGISTER = ImmutableList.of();
+
+  // Format models indexed by file format and object model class
+  private static final Map<Pair<FileFormat, Class<?>>, FormatModel<?, ?>> MODELS =
+      Maps.newConcurrentMap();
+
+  private FormatModelRegistry() {}
+
+  /**
+   * Registers an {@link FormatModel} in this registry.
+   *
+   * <p>The {@link FormatModel} creates readers and writers for a specific combinations of file
+   * format (Parquet, ORC, Avro) and object model (for example: "generic", "spark", "flink", etc.).
+   * Registering custom factories allows integration of new data processing engines for the
+   * supported file formats with Iceberg's file access mechanisms.
+   *
+   * <p>Each factory must be uniquely identified by its combination of file format and object model
+   * name. This uniqueness constraint prevents ambiguity when selecting factories for read and write
+   * operations.
+   *
+   * @param formatModel the factory implementation to register
+   * @throws IllegalArgumentException if a factory is already registered for the combination of
+   *     {@link FormatModel#format()} and {@link FormatModel#type()}
+   */
+  public static void register(FormatModel<?, ?> formatModel) {
+    Pair<FileFormat, Class<?>> key = Pair.of(formatModel.format(), formatModel.type());
+
+    FormatModel<?, ?> existing = MODELS.get(key);
```

Comment:
`register` method uses `get` (line 94)) and `put` (line 104) separately. ConcurrentHahsMap doesn't provide the thread safety. 

`putIfAbsent` is also not a good fit here due to the assertion/override behavior. Maybe just make this method `synchronized` on the `MODELS` object?